### PR TITLE
doc: develop: test: Blackbox explanation

### PR DESCRIPTION
**Pull request by LukaszMrugala**
Documentation file for blackbox tests added.

Need for such a file [has been indicated](https://github.com/zephyrproject-rtos/zephyr/pull/67449#discussion_r1450244086) as we are currently developing new blackbox tests, which require people to be able to understand them more easily.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
Hi, the CI failure has been fixed upstream, if you rebase it should pass.

File: doc/develop/test/blackbox.rst [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+======
+
+Twister blackbox tests are written in python, with auxiliary test data following
+whichever format it was in originally. They are wholly contained in the
+`scripts/tests/twister_blackbox/` directory inside zephyr and prepended with `test_`.
+
+Blackbox tests should not be aware of the internal twister code. Instead, they should
+call twister as user would and check the results.
+
+This README aims to explain the structure of a test file so the reader will be able
```
Comment (golowanow) [inline doc/develop/test/blackbox.rst:17]: now this guide is a part of Zephyr documentation, not just a README as it was.
Would be nice to have  README as well just with few examples of how to run these tests.

**Review by KamilxPaszkiet — DISMISSED** [review]

**Review by golowanow — COMMENTED** [review]
I think this might be a good starter point to document the `balackbox` suite. 
Can it also describe what tests are there and how it executes with its fakes/stubs, etc.
And just few more examples on how to resolve regressions or extend for a new Twister feature.

File: doc/develop/test/blackbox.rst [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+======================
+
+Basics
+======
+
+Twister blackbox tests are written in python, with auxiliary test data following
+whichever format it was in originally. They are wholly contained in the
+`scripts/tests/twister_blackbox/` directory inside zephyr and prepended with `test_`.
+
+Blackbox tests should not be aware of the internal twister code. Instead, they should
```
Comment (hakehuang) [inline doc/develop/test/blackbox.rst:14]: Black Box Testing Techniques usually has below:

<html>
<body>

Decision Table Technique | Decision Table Technique is a systematic approach where various input combinations and their respective system behavior are captured in a tabular form. It is appropriate for the functions that have a logical relationship between two and more than two inputs.
-- | --
Boundary Value Technique | Boundary Value Technique is used to test boundary values, boundary values are those that contain the upper and lower limit of a variable. It tests, while entering boundary value whether the software is producing correct output or not.
State Transition Technique | State Transition Technique is used to capture the behavior of the software application when different input values are given to the same function. This applies to those types of applications that provide the specific number of attempts to access the application.
All-pair Testing Technique | All-pair testing Technique is used to test all the possible discrete combinations of values. This combinational method is used for testing the application that uses checkbox input, radio button input, list box, text box, etc.
Cause-Effect Technique | Cause-Effect Technique underlines the relationship between a given result and all the factors affecting the result.It is based on a collection of requirements.
Equivalence Partitioning Technique | Equivalence partitioning is a technique of software testing in which input data divided into partitions of valid and invalid values, and it is mandatory that all partitions must exhibit the same behavior.
Error Guessing Technique | Error guessing is a technique in which there is no specific method for identifying the error. It is based on the experience of the test analyst, where the tester uses the experience to guess the problematic areas of the software.
Use Case Technique | Use case Technique used to identify the test cases from the beginning to the end of the system as per the usage of the system. By using this technique, the test team creates a test scenario that can exercise the entire software based on the functionality of each function from start to end.


</body>
</html>[Decision Table Technique](https://www.javatpoint.com/decision-table-technique-in-black-box-testing)	Decision Table Technique is a systematic approach where various input combinations and their respective system behavior are captured in a tabular form. It is appropriate for the functions that have a logical relationship between two and more than two inputs.
[Boundary Value Technique](https://www.javatpoint.com/boundary-value-analysis-in-black-box-testing)	Boundary Value Technique is used to test boundary values, boundary values are those that contain the upper and lower limit of a variable. It tests, while entering boundary value whether the software is producing correct output or not.
[State Transition Technique](https://www.javatpoint.com/state-transition-technique-in-black-box-testing)	State Transition Technique is used to capture the behavior of the software application when different input values are given to the same function. This applies to those types of applications that provide the specific number of attempts to access the application.
[All-pair Testing Technique](https://www.javatpoint.com/all-pairs-testing-technique-in-black-box-testing)	All-pair testing Technique is used to test all the possible discrete combinations of values. This combinational method is used for testing the application that uses checkbox input, radio button input, list box, text box, etc.
[Cause-Effect Technique](https://www.javatpoint.com/cause-and-effect-graph-technique-in-black-box-testing)	Cause-Effect Technique underlines the relationship between a given result and all the factors affecting the result.It is based on a collection of requirements.
[Equivalence Partitioning Technique](https://www.javatpoint.com/equivalence-partitioning-technique-in-black-box-testing)	Equivalence partitioning is a technique of software testing in which input data divided into partitions of valid and invalid values, and it is mandatory that all partitions must exhibit the same behavior.
[Error Guessing Technique](https://www.javatpoint.com/error-guessing-technique-in-black-box-testing)	Error guessing is a technique in which there is no specific method for identifying the error. It is based on the experience of the test analyst, where the tester uses the experience to guess the problematic areas of the software.
[Use Case Technique](https://www.javatpoint.com/use-case-technique-in-black-box-testing)	Use case Technique used to identify the test cases from the beginning to the end of the system as per the usage of the system. By using this technique, the test team creates a test scenario that can exercise the entire software based on the functionality of each function from start to end.

what do we plan to use here?

File: doc/develop/test/blackbox.rst [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+This README aims to explain the structure of a test file so the reader will be able
+to understand existing files and create their own. All developers should fix any tests
+they break and create new ones when introducing new features, so this knowledge is
+important for any Twister developer.
+
+Sample test file
+================
+
+.. code:: python
+
```
[the remark this answers is not in the document]
Comment (LukaszMrugala) [reply doc/develop/test/blackbox.rst]: Couldn't get `.rst` to recognise the `:linenos:` command here - switched to `.. include::`, which, coincidentally, improved readability of the `.rst` file.

File: doc/develop/test/blackbox.rst (hunk 2 above)
Comment (LukaszMrugala) [reply doc/develop/test/blackbox.rst:14]: Current approach is not directly based on any blackbox testing framework. Current aim is to cover every Twister flag individually. We do not aim to have a full combinatorial coverage of flag combinations. That would make it most similar to the Cause-Effect Technique from the table. However, applying typical blackbox testing techniques is harder, as practically all Twister flags and setting have scarce descriptions of their desired results - a far cry from a well-defined input-output contracts common to Decision Table Technique.

File: doc/develop/test/blackbox.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
+.. _blackbox:
```
Comment (nashif) [inline doc/develop/test/blackbox.rst:1]: Please move this into a subfolder and make twister specific. _blackbox is very generic of a reference if this is only about twister.

File: doc/develop/test/blackbox.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+======================
```
Comment (nashif) [inline doc/develop/test/blackbox.rst:3]: No need for this

File: doc/develop/test/blackbox.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+Twister blackbox tests
+======================
+
+Basics
+======
```
Comment (nashif) [inline doc/develop/test/blackbox.rst:8]: This should be on a different level

**Comment by nashif** [comment]
why is this still in draft?

File: doc/develop/test/twister/twister_blackbox.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+This guide aims to explain the structure of a test file so the reader will be able
+to understand existing files and create their own. All developers should fix any tests
+they break and create new ones when introducing new features, so this knowledge is
+important for any Twister developer.
+
+Basics
+******
+
+Twister blackbox tests are written in python, with auxiliary test data following
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:14]: I think it is essential to mention `pytest` and refer to its doc pages wherever it is helpful.
And it should be clear how it relates to the pytest Twister plugin https://docs.zephyrproject.org/latest/develop/test/pytest.html

File: doc/develop/test/twister/twister_blackbox.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+whichever format it was in originally. They are wholly contained in the
+`scripts/tests/twister_blackbox/` directory inside zephyr and prepended with `test_`.
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:16]:
```suggestion
:zephyr_file:`scripts/tests/twister_blackbox` directory and prepended with `test_`.
```

File: doc/develop/test/twister/twister_blackbox.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+.. include:: ./sample_blackbox_test.py
+   :literal:
+   :code: python
+   :number-lines:
+
+Comparison with CLI
+*******************
+
+Test above runs the command
+``twister -i --outdir $OUTDIR -T $TEST_DATA/tests -y --level $LEVEL --test-config $TEST_DATA/test_config.yaml -p qemu_x86 -p frdm_k64f`` .
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:33]:
```suggestion
.. code-block:: console

twister -i --outdir $OUTDIR -T $TEST_DATA/tests -y --level $LEVEL --test-config $TEST_DATA/test_config.yaml -p qemu_x86 -p frdm_k64f

```
same in other places

File: doc/develop/test/twister/twister_blackbox.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+.. include:: ./sample_blackbox_test.py
+   :literal:
+   :code: python
+   :number-lines:
+
+Comparison with CLI
+*******************
+
+Test above runs the command
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:32]: and how to run that test above, with what command line ?

File: doc/develop/test/twister/twister_blackbox.rst [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+``twister -i --outdir $OUTDIR -T $TEST_DATA/tests -y --level $LEVEL --test-config $TEST_DATA/test_config.yaml -p qemu_x86 -p frdm_k64f`` .
+The flags are set in the ``args`` variable, which is a list of ``str`` 's [#f1]_ .
+The arguments are then provided to Twister in the first ``with`` segment via ``mock`` library.
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:35]: please put links to related libraries/methods etc. - like in the second part of this guide.

File: doc/develop/test/twister/twister_blackbox.rst [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+Such a test provides us with all the outputs we typically expect of a Twister run. We can
+check the standard output or stderr thanks to ``capfd`` . The three-line block starting with
+``readouterr()`` invocation is commonly used in tests. stdout is written to an ``out`` variable
+and stderr to ``err`` variable. The following lines ensure that the user can still see the outputs
+in their console after running the test.
+
+Beside the standard outputs, we can also investigate the file outputs, normally placed In
+``twister-out`` directories. Most of the time, we will use the ``out_path`` fixture in conjunction
+with ``--outdir`` flag to keep test-generated files in temporary directories. In the example above,
```
Comment (golowanow) [inline doc/develop/test/twister/twister_blackbox.rst:48]: command line examples here

**Review by KamilxPaszkiet — DISMISSED** [review]

File: doc/develop/test/twister/twister_blackbox.rst [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+Fixtures
+========
+
+Blackbox tests use ``pytest`` 's fixtures,
+further reading on which is available
+`here <https://docs.pytest.org/en/6.2.x/fixture.html>`__.
+
+If you would like to add your own fixtures,
+consider whether they will be used in just one test file, or in many.
```
Comment (kartben) [inline doc/develop/test/twister/twister_blackbox.rst:97]: please check line wrapping settings and ensure proper formatting of the file, as this here seems pretty aggressive :) You can have 100 characters per line.

**Review by kartben — CHANGES_REQUESTED** [review]

**Review by kartben — APPROVED** [review]

**Review by nashif — APPROVED** [review]
